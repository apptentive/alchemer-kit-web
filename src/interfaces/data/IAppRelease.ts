export interface IAppRelease {
  type: string;
  version?: string | null;
  build_number?: number | null;
  sdk_version: string;
}
