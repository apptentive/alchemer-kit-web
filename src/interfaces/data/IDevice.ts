export interface IDevice {
  [key: string]: any;
  browser_height: number;
  browser_width: number;
  cookie_enabled: boolean;
  browser_languages: readonly string[];
  locale_language_code: string;
  locale_raw: string;
  user_agent: string;
  browser_vendor: string;
  screen_width: number;
  screen_height: number;
  orientation_angle: number;
  orientation_type: OrientationType;
  uuid: string;
  custom_data: object;
}

export enum OsNameEnum {
  // TODO: uncomment when os_version for windows will be implemented
  // WINDOWS_10 = 'Windows 10',
  // WINDOWS_8_1 = 'Windows 8.1',
  // WINDOWS_8 = 'Windows 8',
  // WINDOWS_7 = 'Windows 7',
  // WINDOWS_VISTA = 'Windows Vista',
  // WINDOWS_XP = 'Windows XP',
  // WINDOWS_SERVER_2003 = 'Windows Server 2003',
  // WINDOWS_2000 = 'Windows 2000',
  // WINDOWS_CE = 'Windows CE',
  // WINDOWS_98 = 'Windows 98',
  // WINDOWS_ME = 'Windows ME',
  // WINDOWS_95 = 'Windows 95',
  WINDOWS = 'Windows',
  MAC_OS = 'MacOS',
  LINUX = 'Linux',
  ANDROID = 'Android',
  IOS = 'iOS',
}
