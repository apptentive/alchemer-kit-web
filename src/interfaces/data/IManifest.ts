import { IInteraction } from '../manifest/IInteraction';
import { ITargetedEvents } from '../manifest/ITargetedEvents';

export interface IManifest {
  interactions: IInteraction[];
  targets: ITargetedEvents;
}
