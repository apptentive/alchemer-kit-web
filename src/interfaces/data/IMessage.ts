export interface IMessage {
  body: string;
  custom_data?: {
    [key: string]: any;
  };
}
