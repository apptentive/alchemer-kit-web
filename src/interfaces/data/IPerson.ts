export interface IPerson {
  id: string;
  name?: string;
  email?: string;
  phone_number?: string;
  address?: string;
  birthday?: string;
  custom_data?: {
    [key: string]: number | string | boolean;
  };
}
