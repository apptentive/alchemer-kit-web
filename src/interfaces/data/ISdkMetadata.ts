export interface ISdkMetadata {
  programming_language: string;
  author_name: string;
  author_email?: string;
  platform: string;
  distribution: string;
  distribution_version?: string;
  version: string;
}
