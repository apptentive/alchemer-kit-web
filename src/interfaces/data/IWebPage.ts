export interface IWebPage {
  name: string;
  path: string;
  search: string;
  title: string;
  url: string;
}
