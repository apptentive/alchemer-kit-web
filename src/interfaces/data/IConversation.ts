import { IAppRelease } from './IAppRelease';
import { IPerson } from './IPerson';
import { ISdkMetadata } from './ISdkMetadata';

export interface IConversation {
  state: 'new';
  id: string;
  device_id: string;
  person_id: string;
  sdk: ISdkMetadata;
  app_release?: IAppRelease;
  token: string;
  person?: IPerson;
}
