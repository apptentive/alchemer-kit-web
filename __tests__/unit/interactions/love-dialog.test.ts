import ApptentiveLoveDialog from '../../../src/interactions/love-dialog';
import ApptentiveBaseBuilder from '../../mocks/builders/ApptentiveBaseBuilder';

import { _appId, _apiToken, _domNode, _domNodeId, _host, _interactionOptions } from '../../mocks/data/shared-constants';

const loveDialogConfig = {
  id: '55bae8ef35878bc3c600057c',
  type: 'EnjoymentDialog',
  version: 1,
  configuration: {
    title: 'Do you love app?',
    yes_text: 'Yes!',
    no_text: 'No',
    show_dismiss_button: false,
  },
} as any;

beforeEach(() => {
  const div = document.createElement('div');
  div.id = _domNodeId;
  div.innerHTML = '';
  document.body.append(div);
});

afterEach(() => {
  document.body.innerHTML = '';
});

test('#constructor: has no side effect if no config is passed', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();

  // @ts-expect-error test for passing invalid config
  const loveDialog = new ApptentiveLoveDialog(undefined, sdk);
  loveDialog.render();
  expect(document.querySelector(_domNode)!.innerHTML).toBe('');
});

test('#render: should render a DOM element on the page', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig, sdk, _interactionOptions);
  loveDialog.render();

  expect(document.querySelector(_domNode)!.innerHTML).toMatchSnapshot();
});

test('#render: should render properly if configured DOM node does not exist', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig, sdk, { domNode: '#existentialElement' });
  loveDialog.render();

  expect(document.querySelector('body > apptentive-love-dialog')).not.toBeNull();
});

test('#render: can be built as a static method', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();

  ApptentiveLoveDialog.display(loveDialogConfig, sdk, _interactionOptions);
  expect(document.querySelector(_domNode)!.innerHTML).toMatchSnapshot();
});

test('#render: does not throw if no parameters are passed', () => {
  // @ts-expect-error test validating no error when missing parameters
  expect(() => ApptentiveLoveDialog.display()).not.toThrow();
});

test('#render: renders only once', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig, sdk, _interactionOptions);
  loveDialog.render();
  loveDialog.render();

  expect(document.querySelectorAll('apptentive-love-dialog').length).toBe(1);
});

test('#cancel: should not throw an error', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig, sdk, _interactionOptions);
  loveDialog.render();

  expect(() => {
    loveDialog.cancel(new Event('cancel'));
  }).not.toThrow();
});

test('#cancel: should handle error if no sdk is available', () => {
  // @ts-expect-error test for missing sdk and options
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig);

  expect(() => loveDialog.cancel(new Event('cancel'))).not.toThrow();
});

test('#dismiss: should not throw an error', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig, sdk, _interactionOptions);
  loveDialog.render();

  expect(() => {
    loveDialog.dismiss(new Event('dismiss'));
  }).not.toThrow();
});

test('#dismiss: should handle error if no sdk is available', () => {
  // @ts-expect-error test for missing sdk parameter
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig);

  expect(() => loveDialog.dismiss(new Event('dismiss'))).not.toThrow();
});

test('#dismiss: should remove the interaction from the DOM', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig, sdk, _interactionOptions);
  loveDialog.render();

  expect(document.querySelectorAll('apptentive-love-dialog').length).toBe(1);

  const dismiss = document.querySelector<HTMLButtonElement>('.close-love-dialog')!;
  dismiss.click();

  expect(document.querySelectorAll('apptentive-love-dialog').length).toBe(0);
});

test('#cancel: should engage a close and cancel event', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig, sdk, _interactionOptions);
  loveDialog.render();

  const apiSpy = jest.spyOn(loveDialog.sdk, 'engage');

  const cancel = document.querySelector<HTMLButtonElement>('.close-love-dialog')!;
  cancel.click();

  expect(apiSpy.mock.calls[0][0]).toBe('com.apptentive#EnjoymentDialog#cancel');
  expect(apiSpy.mock.calls[0][1]!.id).toBe('55bae8ef35878bc3c600057c');

  apiSpy.mockRestore();
});

test('Yes Button: should remove the interaction from the DOM', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig, sdk, _interactionOptions);
  loveDialog.render();

  expect(document.querySelectorAll('apptentive-love-dialog').length).toBe(1);

  const dismiss = document.querySelector<HTMLButtonElement>(
    '.apptentive-love-dialog-action.apptentive-love-dialog-yes'
  )!;
  dismiss.click();

  expect(document.querySelectorAll('apptentive-love-dialog').length).toBe(0);
});

test('Yes Button: should engage a close event', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig, sdk, _interactionOptions);
  loveDialog.render();

  const apiSpy = jest.spyOn(loveDialog.sdk, 'engage');
  const dismiss = document.querySelector<HTMLButtonElement>(
    '.apptentive-love-dialog-action.apptentive-love-dialog-yes'
  )!;
  dismiss.click();

  expect(apiSpy.mock.calls[0][0]).toBe('com.apptentive#EnjoymentDialog#yes');
  expect(apiSpy.mock.calls[0][1]!.id).toBe('55bae8ef35878bc3c600057c');
  expect(apiSpy.mock.calls[0][1]!.label).toBe('Yes!');

  apiSpy.mockRestore();
});

test('No Button: should remove the interaction from the DOM', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig, sdk, _interactionOptions);
  loveDialog.render();

  expect(document.querySelectorAll('apptentive-love-dialog').length).toBe(1);

  const dismiss = document.querySelector<HTMLButtonElement>(
    '.apptentive-love-dialog-action.apptentive-love-dialog-no'
  )!;
  dismiss.click();

  expect(document.querySelectorAll('apptentive-love-dialog').length).toBe(0);
});

test('No Button: should engage a close event', () => {
  const { sdk } = new ApptentiveBaseBuilder().useCachedManifest().build();
  const loveDialog = new ApptentiveLoveDialog(loveDialogConfig, sdk, _interactionOptions);
  loveDialog.render();

  const apiSpy = jest.spyOn(loveDialog.sdk, 'engage');

  const dismiss = document.querySelector<HTMLButtonElement>(
    '.apptentive-love-dialog-action.apptentive-love-dialog-no'
  )!;
  dismiss.click();

  expect(apiSpy.mock.calls[0][0]).toBe('com.apptentive#EnjoymentDialog#no');
  expect(apiSpy.mock.calls[0][1]!.id).toBe('55bae8ef35878bc3c600057c');
  expect(apiSpy.mock.calls[0][1]!.label).toBe('No');

  apiSpy.mockRestore();
});
