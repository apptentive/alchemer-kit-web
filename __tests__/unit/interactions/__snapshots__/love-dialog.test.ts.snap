// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#render: can be built as a static method 1`] = `"<apptentive-love-dialog id="apptentive-love-dialog-55bae8ef35878bc3c600057c" class="fixed corner"><div class="apptentive-love-dialog-content"><div class="close-love-dialog">✕</div><div class="apptentive-love-dialog-title"><h1>Do you love app?</h1></div><div class="apptentive-love-dialog-actions"><button class="apptentive-love-dialog-action apptentive-love-dialog-no">No</button><button class="apptentive-love-dialog-action apptentive-love-dialog-yes">Yes!</button></div></div></apptentive-love-dialog>"`;

exports[`#render: should render a DOM element on the page 1`] = `"<apptentive-love-dialog id="apptentive-love-dialog-55bae8ef35878bc3c600057c" class="fixed corner"><div class="apptentive-love-dialog-content"><div class="close-love-dialog">✕</div><div class="apptentive-love-dialog-title"><h1>Do you love app?</h1></div><div class="apptentive-love-dialog-actions"><button class="apptentive-love-dialog-action apptentive-love-dialog-no">No</button><button class="apptentive-love-dialog-action apptentive-love-dialog-yes">Yes!</button></div></div></apptentive-love-dialog>"`;
